### PR TITLE
Allow implicit conversion of `PredicateBuilder` to delegates

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

Contributes to #1329

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: bench/Polly.Core.Benchmarks/PredicateBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            .HandleResult(r => r.StatusCode == HttpStatusCode.InternalServerError)
+            .Handle<HttpRequestException>()
+            .Handle<InvalidOperationException>(e => false)
+    };
+
+    [Benchmark(Baseline = true)]
+    public ValueTask<bool> Predicate_SwitchExpression()
+    {
+        return _delegate.ShouldHandle(_args);
+
```
Comment (martincostello) [inline bench/Polly.Core.Benchmarks/PredicateBenchmark.cs:39]:
```suggestion
```

File: src/Polly.Core/PredicateBuilder.TResult.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return Add(outcome => outcome.Exception?.InnerException is TException innerException && predicate(innerException));
+    }
+
+    /// <summary>
+    /// Adds a predicate for handling results.
+    /// </summary>
+    /// <param name="predicate">The predicate function to use for handling the result.</param>
+    /// <returns>The same instance of the <see cref="PredicateBuilder{TResult}"/> for chaining.</returns>
+    public PredicateBuilder<TResult> HandleResult(Func<TResult, bool> predicate)
+        => Add(outcome => outcome.TryGetResult(out var result) && predicate(result!));
```
Comment (martincostello) [inline src/Polly.Core/PredicateBuilder.TResult.cs:69]: Can we annotate `TryGetResult` so the compiler knows `result` won't be `null` if it returns `true`? Then we won't need the `!`.

File: src/Polly.Core/PredicateBuilder.TResult.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    /// <summary>
+    /// Builds the predicate.
+    /// </summary>
+    /// <returns>An instance of predicate delegate.</returns>
+    /// <exception cref="InvalidOperationException">Thrown when no predicates were configured using this builder.</exception>
+    /// <remarks>
+    /// The returned predicate will return <see langword="true"/> if any of the configured predicates return <see langword="true"/>.
+    /// Please be aware of performance penalty if you register too many predicates with this builder. In such case, it's better to create your own predicate
```
Comment (martincostello) [inline src/Polly.Core/PredicateBuilder.TResult.cs:91]: Here and other presences.

```suggestion
    /// Please be aware of the performance penalty if you register too many predicates with this builder. In such case, it's better to create your own predicate
```

File: src/Polly.Core/PredicateBuilder.TResult.cs (hunk 2 above)
Comment (martintmk) [reply src/Polly.Core/PredicateBuilder.TResult.cs:69]: The problem is that we should also allow `nulls` there.

You can do `Outcome.FromResult<object?>(null)`. It's perfectly valid result.

**Review by martincostello — APPROVED** [review]

**Pull request merged**